airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -136,34 +147,92 @@ def _create_prepared_request(
 
         return prepared_request
 
-    def _send_with_retry(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+    def _max_retries(self) -> int:
+        max_retries = self._DEFAULT_MAX_RETRY
+        if self._disable_retries:
+            max_retries = 0
+        elif hasattr(self._error_handler, "max_retries") and self._error_handler.max_retries is not None:
```

Comment:
This condition seems very odd. Why do we have this here? Should we make it that everything relies on the interface `ErrorHandler`? 

The same comment applies to `max_time` and `retry_factor`

Reply:
Yeah I recognize this is pretty weird.

Maybe I'm being too limiting, but the issue is that the existing low-code error handlers (e.g. `DefaultErrorHandler`) includes max retries and max_time, and the max_retries is actually specified by the connector dev in the manifest. The existing low-code backoff handlers (e.g. `ExponentialBackoffStrategy`) may include retry factor.

In the Python CDK, the `DefaultBackoffStrategy` includes max_retries and max_time.

In other words, these properties aren't enforced by the common interface. In order to minimize impact to existing low-code components (specifically because the error handler schema enables connector dev to specify a max_retry count), and because multiple backoff strategies can be specified for low-code. I opted to try to cater to both low-code and python implementations.

Let me know if you think it would be preferred to align more closely.

Reply:
This somewhat makes sense but I'm still sceptic because:
* It feels like we are creating this brand new shiny class and it's already legacy because we purposefully decide not to respect the interfaces of our classes
* If I'm doing some development on the declarative package and changing the `max_retries` interface, now this will be broken without having some clear dependencies (the typing does not reflect that)
* It create weird cases where if I define an error_handler AND a backoff strategy that both have a max_retries, then the case is not clear for the user of this class

It feels like an easy problem to solve though: `HttpRequester` should instantiate proper implementations of `airbyte_cdk.sources.streams.http.error_handlers.BackoffStrategy`. What is preventing us from doing that?

Reply:
@maxi297 Alright I agree with you.

I've update the ErrorHandler interface to enforce max_retries, max_time and aligned the `HttpStatusErrorHandler` with the updated interface.

Now all ErrorHandler implementations will have max_retries and max_time.

As for `factor`, that is still only available included for the `ExponentialBackoffStrategy` (or any custom backoff strategy that wants to implement a custom backoff strategy, so I kept the `hasattr` check for that.

Updated accordingly.